### PR TITLE
[#16] [Backend] As a user, I can see the global feeds from the home screen

### LINKED ISSUE #16: As a user, I can see the global feeds from the home screen
## Why
The users will be able to see the global feeds of articles in the `Home` screen regardless whether they are logged in or not with data provided from server.

## Acceptance Criteria
- [ ] Implement the API for getting global feeds data via this [endpoint](https://github.com/gothinkster/realworld/tree/master/api#list-articles).
- [ ] Setup the default limit articles per page as `25`.
- [ ] Setup pagination support for this endpoint.
- [ ] Create a usecase for getting global feeds for later usage in integrate task.
- [ ] Parse the response into the corresponding models, here is an example response for the list of articles:
```
{
  "articles":[{
    "slug": "how-to-train-your-dragon",
    "title": "How to train your dragon",
    "description": "Ever wonder how?",
    "body": "It takes a Jacobian",
    "tagList": ["dragons", "training"],
    "createdAt": "2016-02-18T03:22:56.637Z",
    "updatedAt": "2016-02-18T03:48:35.824Z",
    "favorited": false,
    "favoritesCount": 0,
    "author": {
      "username": "jake",
      "bio": "I work at statefarm",
      "image": "https://i.stack.imgur.com/xHWG8.jpg",
      "following": false
    }
  }, {
    "slug": "how-to-train-your-dragon-2",
    "title": "How to train your dragon 2",
    "description": "So toothless",
    "body": "It a dragon",
    "tagList": ["dragons", "training"],
    "createdAt": "2016-02-18T03:22:56.637Z",
    "updatedAt": "2016-02-18T03:48:35.824Z",
    "favorited": false,
    "favoritesCount": 0,
    "author": {
      "username": "jake",
      "bio": "I work at statefarm",
      "image": "https://i.stack.imgur.com/xHWG8.jpg",
      "following": false
    }
  }],
  "articlesCount": 2
}
```

### PR DESCRIPTION
**Pull request by markgravity**
Resolved #16 

## What happened

Create Article List API

## Insight
- Add `R.swift` into Tests target
- Add `ArticleRepository`
- Add `ListArticlesUseCase`
- Write unit test for ArticleRepository` & `ListArticlesUseCase `

## Proof Of Work

```
*** Overview *** 
URL https://realworld-temp-api.herokuapp.com/api/articles
Method GET
Response Code 200
Request Start Time Sep 9 2021 - 11:30:07
Duration 1s

*** Request Header *** 
Accept-Encoding br;q=1.0, gzip;q=0.9, deflate;q=0.8
User-Agent Nimble Medium/1.0.0 (co.nimblehq.nimble-medium; build:1; iOS 14.5.0) Alamofire/5.4.3
Accept-Language en;q=1.0


*** Request Body *** 
-

*** Response Header *** 
Access-Control-Allow-Origin *
Connection keep-alive
Date Thu, 09 Sep 2021 04:30:09 GMT
Content-Length 694
Content-Type application/json; charset=utf-8
X-Powered-By Express
Server Cowboy
Etag W/"2b6-EB329SMWH8gKzKe4EaSoYn/G30I"
Via 1.1 vegur


*** Response Body *** 
{
  "articlesCount" : 1,
  "articles" : [
    {
      "slug" : "Welcome-to-RealWorld-project",
      "tagList" : [
        "introduction",
        "welcome"
      ],
      "author" : {
        "bio" : null,
        "username" : "Gerome",
        "image" : "https://realworld-temp-api.herokuapp.com/images/demo-avatar.png"
      },
      "comments" : [

      ],
      "favoritesCount" : 1,
      "title" : "Welcome to RealWorld project",
      "favorited" : false,
      "updatedAt" : "2021-09-05T22:23:59.819Z",
      "description" : "Exemplary fullstack Medium.com clone powered by React, Angular, Node, Django, and many more",
      "body" : "See how the exact same Medium.com clone (called Conduit) is built using different frontends and backends. Yes, you can mix and match them, because they all adhere to the same API spec",
      "createdAt" : "2021-09-05T22:23:59.819Z"
    }
  ]
}

------------------------------------------------------------------------
------------------------------------------------------------------------
------------------------------------------------------------------------
```

Rxswift Debug
```
2021-09-09 11:30:09.577: App.swift:27 (init()) -> Event next([NimbleMedium.CodableArticle(slug: "Welcome-to-RealWorld-project", title: "Welcome to RealWorld project", description: "Exemplary fullstack Medium.com clone powered by React, Angular, Node, Django, and many more", body: "See how the exact same Medium.com clone (called Conduit) is built using different frontends and backends. Yes, you can mix and match them, because they all adhere to the same API spec", tagList: ["introduction", "welcome"], createdAt: 2021-09-05 22:23:59 +0000, updatedAt: 2021-09-05 22:23:59 +0000, favorited: false, favoritesCount: 1, apiAuthor: NimbleMedium.CodableProfile(username: "Gerome", bio: nil, image: Optional("https://realworld-temp-api.herokuapp.com/images/demo-avatar.png"), _following: NimbleMedium.DecodableDefault.Wrapper<NimbleMedium.DecodableDefault.Sources.False>(wrappedValue: false)))])
```

### CODE REVIEW

File: .github/workflows/lint_and_unit_test.yml [hunk 1]
```diff
@@ -55,4 +55,14 @@ jobs:
...
+
+    - name: Generate code coverage file
+      run: slather
+
+    - name: Upload code coverage to Codecov
+      uses: codecov/codecov-action@v2
+      with:
+        files: ./cobertura.xml
+        fail_ci_if_error: false
+        verbose: true
```
Comment (minhnimble) [inline .github/workflows/lint_and_unit_test.yml:68]: Please help create a separate PR for the implement for code coverage since it is out of scope of the current ticket.

File: NimbleMedium/Sources/Constants/Constants+API.swift [hunk 2]
```diff
@@ -9,5 +9,5 @@
 extension Constants.API {
 
     // For fixing performance issues and the presence of NSFW content on the current API, we encourage you to use the new temporary API: https://realworld-temp-api.herokuapp.com/api Please keep in mind we'll soon move back to the current API: https://conduit.productionready.io/api
-    static let baseURL: String = "https://conduit.productionready.io/api"
+    static let baseURL: String = "https://realworld-temp-api.herokuapp.com/api"
```
Comment (minhnimble) [inline NimbleMedium/Sources/Constants/Constants+API.swift:12]: please don't commit this endpoint up yet, this is for local testing purpose only

File: NimbleMedium/Sources/Data/Repositories/Article/ArticleRepository.swift [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    private let networkAPI: NetworkAPIProtocol
+
+    init(networkAPI: NetworkAPIProtocol) {
+        self.networkAPI = networkAPI
+    }
+
+    func listArticles(
+        tag: String?,
+        author: String?,
+        favorited: String?,
```
Comment (minhnimble) [inline NimbleMedium/Sources/Data/Repositories/Article/ArticleRepository.swift:21]: do we need to provide these attributes? I think these are out of scope of this ticket 🤔

File: NimbleMedium/Sources/Supports/Extensions/Foundation/DateFormatter+ISO8601Full.swift [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+//  Created by Mark G on 06/09/2021.
+//
+
+import Foundation
+
+extension DateFormatter {
+
+    static let iso8601Full: DateFormatter = {
+        let formatter = DateFormatter()
+        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSZZZZZ"
```
Comment (minhnimble) [inline NimbleMedium/Sources/Supports/Extensions/Foundation/DateFormatter+ISO8601Full.swift:14]: please put this date format in constant file for organizing purpose 🙏

File: NimbleMedium/Sources/Supports/Helpers/Foundation/DecodableDefault.swift [hunk 5]
```diff
@@ -0,0 +1,98 @@
+//
+//  Decodable.swift
+//
+//  Copyright © 2021 Leandro Nunes Fantinatto. All rights reserved.
+//  Ref: https://github.com/gonzalezreal/DefaultCodable
+//  Ref: https://github.com/JohnSundell/Codextended
+//
```
Comment (minhnimble) [inline NimbleMedium/Sources/Supports/Helpers/Foundation/DecodableDefault.swift:7]: you copy this from somewhere? 😁 Does it has a Pod or SPM for us to use instead of including the file directly like this? 🤔

File: NimbleMediumTests/Sources/Utilities/Data+Decode.swift [hunk 6]
```diff
@@ -0,0 +1,37 @@
+//
+//  Data+Decode.swift
+//  MVVMRxSwiftDemoTests
```
Comment (minhnimble) [inline NimbleMediumTests/Sources/Utilities/Data+Decode.swift:3]: this is not our project, please update to our project name correspondingly 🙏

File: NimbleMediumTests/Sources/Utilities/Data+Decode.swift [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+//
+//  Created by Bliss on 1/7/21.
+//
+// swiftlint:disable force_try
```
Comment (minhnimble) [inline NimbleMediumTests/Sources/Utilities/Data+Decode.swift:7]: can we fix the code to remove this SwiftLint check? 🙏

File: NimbleMediumTests/Sources/Utilities/String+Data.swift [hunk 8]
```diff
@@ -0,0 +1,15 @@
+//
+//  String+Data.swift
+//  MVVMRxSwiftDemoTests
```
Comment (minhnimble) [inline NimbleMediumTests/Sources/Utilities/String+Data.swift:3]: same here

File: NimbleMediumTests/Sources/Mocks/NetworkAPIProtocolMock.swift [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    func setPerformRequestForReturnValue<T: Decodable>(_ value: Single<T>) {
+        performRequestForReturnValues["\(T.self)"] = value
+    }
+
+    func getPerformRequestForReturnValue<T: Decodable>(for type: T.Type) -> Single<T> {
+        performRequestForReturnValues["\(T.self)"] as! Single<T>
+    }
+
```
Comment (minhnimble) [inline NimbleMediumTests/Sources/Mocks/NetworkAPIProtocolMock.swift:38]: this break line is redundant, can remove

File: NimbleMediumTests/Sources/Dummy/Data/Models/CodableArticle+Dummy.swift [hunk 10]
```diff
@@ -0,0 +1,40 @@
+//
+//  APISurvey+Dummy.swift
+//  MVVMRxSwiftDemoTests
```
Comment (minhnimble) [inline NimbleMediumTests/Sources/Dummy/Data/Models/CodableArticle+Dummy.swift:3]: same here

**Comment by markgravity** [comment]
sd

File: NimbleMedium/Sources/Data/Repositories/Article/ArticleRepository.swift (hunk 3 above)
Comment (markgravity) [reply NimbleMedium/Sources/Data/Repositories/Article/ArticleRepository.swift:21]: I think it's fine, we're trying to implement this [api](https://github.com/gothinkster/realworld/tree/main/api#list-articles) right?
So if we added all available parameters so that we dont need to another PR when we need the rest 🤔

File: NimbleMediumTests/Sources/Utilities/Data+Decode.swift (hunk 7 above)
Comment (markgravity) [reply NimbleMediumTests/Sources/Utilities/Data+Decode.swift:7]: @minhnimble It's from our ios-template, so I guess we can improve it later in our ios-template 🙏

File: NimbleMedium/Sources/Data/Repositories/Article/ArticleRepository.swift (hunk 3 above)
Comment (minhnimble) [reply NimbleMedium/Sources/Data/Repositories/Article/ArticleRepository.swift:21]: For now I think it is still okay to keep but you are doing things out of the scope, please try not to do it next time and focus on the requirements of the tickets so that we can complete a ticket faster. 🙏

File: NimbleMediumTests/Sources/Utilities/Data+Decode.swift (hunk 7 above)
Comment (minhnimble) [reply NimbleMediumTests/Sources/Utilities/Data+Decode.swift]: yup you are right 👍

File: NimbleMediumTests/Sources/Specs/Data/Repositories/ArticleRepositorySpec.swift [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+                    .map {
+                        $0.compactMap { $0 as? CodableArticle }
+                    }
+                    .bind(to: outputArticles)
+                    .disposed(by: disposeBag)
+                }
+
+                it("returns correct articles") {
+                    expect(outputArticles.events.first?.value.element) == inputResponse.articles
+                }
```
Comment (minhnimble) [inline NimbleMediumTests/Sources/Specs/Data/Repositories/ArticleRepositorySpec.swift:58]: I think we are missing a case when the api failed as well. Also for this `it` we need a `when` the api is successful message also

File: NimbleMediumTests/Sources/Specs/Domain/UseCases/ListArticlesUseCaseSpec.swift [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+                    .map {
+                        $0.compactMap { $0 as? CodableArticle }
+                    }
+                    .bind(to: outputArticles)
+                    .disposed(by: disposeBag)
+                }
+
+                it("returns correct articles") {
+                    expect(outputArticles.events.first?.value.element) == inputArticles
+                }
```
Comment (minhnimble) [inline NimbleMediumTests/Sources/Specs/Domain/UseCases/ListArticlesUseCaseSpec.swift:59]: same as above, we are missing a test case for api failed as well.

File: NimbleMedium/Sources/Data/NetworkAPI/RequestConfigurations/ArticleRequestConfiguration.swift [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+            if let limit = limit {
+                parameters["limit"] = limit
+            }
+
+            if let offset = offset {
+                parameters["offset"] = offset
+            }
+
+            return parameters
```
Comment (suho) [inline NimbleMedium/Sources/Data/NetworkAPI/RequestConfigurations/ArticleRequestConfiguration.swift:70]:
```suggestion
            let parameters: Parameters = [
            	"tag": tag,
            	"author": author,
            	"favorited": favorited,
            	"limit": limit,
            	"offset": offset,
            ]
            return parameters.compactMapValues { $0 }
```

File: NimbleMedium/Sources/Data/NetworkAPI/RequestConfigurations/ArticleRequestConfiguration.swift [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+        }
+    }
+
+    var encoding: ParameterEncoding { URLEncoding.queryString }
+
+    var headers: HTTPHeaders? { nil }
+
+    var interceptor: RequestInterceptor? { nil }
```
Comment (suho) [inline NimbleMedium/Sources/Data/NetworkAPI/RequestConfigurations/ArticleRequestConfiguration.swift:78]: I think we can create a default value for these properties and remove it here

```suggestion
```

```swift
extension RequestConfiguration {

    var headers: HTTPHeaders? { nil }

    var interceptor: RequestInterceptor? { nil }
}
```

File: NimbleMediumTests/Sources/Dummy/Data/Models/APIArticleResponse+Dummy.swift [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+//  Created by Mark G on 08/09/2021.
+//
+
+@testable import NimbleMedium
+
+extension APIArticleResponse {
+
+    static let dummy: APIArticleResponse = {
+        """
+        {
```
Comment (suho) [inline NimbleMediumTests/Sources/Dummy/Data/Models/APIArticleResponse+Dummy.swift:14]: I would prefer to create a JSON file (for example `articles.json`) then decoded it.

File: NimbleMediumTests/Sources/Mocks/NetworkAPIProtocolMock.swift [hunk 16]
```diff
@@ -0,0 +1,38 @@
+//
```
Comment (suho) [inline NimbleMediumTests/Sources/Mocks/NetworkAPIProtocolMock.swift:1]: I think we can put this file into the folder `Data` based on its layer.
-> `NimbleMediumTests/Sources/Mocks/Data/NetworkAPIProtocolMock.swift`

File: NimbleMediumTests/Sources/Utilities/Data+Decode.swift [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+//
+//  swiftlint:disable force_try
+
+import Foundation
+
+@testable import NimbleMedium
+
+extension Data {
+
+    func decode<T: Decodable>(with type: T.Type) -> T {
```
Comment (suho) [inline NimbleMediumTests/Sources/Utilities/Data+Decode.swift:15]: I think we don't need to this function

File: NimbleMediumTests/Sources/Utilities/Data+Decode.swift [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    func decoded<T: Decodable>() -> T {
+        let decoder = JSONDecoder()
+        decoder.keyDecodingStrategy = .convertFromSnakeCase
+        decoder.dateDecodingStrategy = .formatted(.iso8601Full)
+        return try! decoder.decode(T.self, from: self)
+    }
+
+    func convertToDictionary() -> [String: Any]? {
```
Comment (suho) [inline NimbleMediumTests/Sources/Utilities/Data+Decode.swift:29]: I think we don't need to this function

File: NimbleMedium/Sources/Data/NetworkAPI/Models/APIArticleResponse.swift [hunk 19]
```diff
@@ -0,0 +1,14 @@
+//
+//  APIArticleResponse.swift
+//  NimbleMedium
+//
+//  Created by Mark G on 06/09/2021.
+//
+
+import Foundation
+
+struct APIArticleResponse: Codable, Equatable {
```
Comment (suho) [inline NimbleMedium/Sources/Data/NetworkAPI/Models/APIArticleResponse.swift:10]: Please use `Decodable` instead
```suggestion
struct APIArticleResponse: Decodable, Equatable {
```

File: NimbleMedium/Sources/Data/NetworkAPI/Models/CodableArticle.swift [hunk 20]
```diff
@@ -0,0 +1,31 @@
+//
+//  APIArticle.swift
+//  NimbleMedium
+//
+//  Created by Mark G on 06/09/2021.
+//
+
+import Foundation
+
+struct CodableArticle: Article, Codable, Equatable {
```
Comment (suho) [inline NimbleMedium/Sources/Data/NetworkAPI/Models/CodableArticle.swift:10]: Please use `Decodable` instead

File: NimbleMedium/Sources/Data/NetworkAPI/Models/CodableProfile.swift [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+//  CodableProfile.swift
+//  NimbleMedium
+//
+//  Created by Mark G on 06/09/2021.
+//
+
+import Foundation
+import DefaultCodable
+
+struct CodableProfile: Profile, Codable, Equatable {
```
Comment (suho) [inline NimbleMedium/Sources/Data/NetworkAPI/Models/CodableProfile.swift:11]: Please use `Decodable` instead

File: NimbleMediumTests/Sources/Mocks/Data/NetworkAPIProtocolMock.swift [hunk 22]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    func setPerformRequestForReturnValue<T: Decodable>(_ value: Single<T>) {
+        performRequestForReturnValues["\(T.self)"] = value
+    }
+
+    func getPerformRequestForReturnValue<T: Decodable>(for type: T.Type) -> Single<T> {
+        performRequestForReturnValues["\(T.self)"] as! Single<T>
+    }
+}
```
Comment (minhnimble) [inline NimbleMediumTests/Sources/Mocks/Data/NetworkAPIProtocolMock.swift:38]: why don't you use Sourcery for this?

File: NimbleMedium/Sources/Domain/UseCases/ListArticlesUseCase.swift [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+            favorited: favorited,
+            limit: limit,
+            offset: offset
+        )
+    }
+}
+
+extension ListArticlesUseCaseProtocol {
+
+    // Create default value
```
Comment (minhnimble) [inline NimbleMedium/Sources/Domain/UseCases/ListArticlesUseCase.swift:50]: Do you mean this is a TODO?

File: NimbleMediumTests/Sources/Utilities/Data+Decode.swift [hunk 24]
```diff
@@ -0,0 +1,20 @@
+//
+//  Data+Decode.swift
+//  NimbleMedium
```
Comment (minhnimble) [inline NimbleMediumTests/Sources/Utilities/Data+Decode.swift:3]: should be `NimbleMediumTests` right? 😅

**Review by minhnimble — APPROVED** [review]
LGTM 🚀

**Review by suho — APPROVED** [review]
